ostree-pulp.sh

```diff
@@ -0,0 +1,405 @@
+    fi
+    COMPOSE_ID=$(jq -r '.[0].body.build_id' "$COMPOSE_START")
+
+    # Wait for the compose to finish.
+    greenprint "⏱ Waiting for compose to finish: ${COMPOSE_ID}"
+    while true; do
+        sudo composer-cli --json compose info "${COMPOSE_ID}" | tee "$COMPOSE_INFO" > /dev/null
+        COMPOSE_STATUS=$(jq -r '.[0].body.queue_status' "$COMPOSE_INFO")
+
+        # Is the compose finished?
+        if [[ $COMPOSE_STATUS != RUNNING ]] && [[ $COMPOSE_STATUS != WAITING ]]; then
+            break
+        fi
+
+        # Wait 30 seconds and try again.
+        sleep 5
+    done
+
+    # Capture the compose logs from osbuild.
+    greenprint "💬 Getting compose log and metadata"
+    get_compose_log "$COMPOSE_ID"
+    get_compose_metadata "$COMPOSE_ID"
+
+    # Kill the journal monitor immediately and remove the trap
+    sudo pkill -P ${WORKER_JOURNAL_PID}
+    trap - EXIT
+
+    # Did the compose finish with success?
+    if [[ $COMPOSE_STATUS != FINISHED ]]; then
+        redprint "Something went wrong with the compose. 😢"
+        exit 1
+    fi
+}
+
+# Wait for the ssh server up to be.
+wait_for_ssh_up () {
+    SSH_STATUS=$(sudo ssh "${SSH_OPTIONS[@]}" -i "${SSH_KEY}" "admin@${1}" '/bin/bash -c "echo -n READY"')
+    if [[ $SSH_STATUS == READY ]]; then
+        echo 1
+    else
+        echo 0
+    fi
+}
+
+# Clean up our mess.
+clean_up () {
+    greenprint "🧼 Cleaning up"
+    sudo virsh destroy "${IMAGE_KEY}"
+    if [[ $ARCH == aarch64 ]]; then
+        sudo virsh undefine "${IMAGE_KEY}" --nvram
+    else
+        sudo virsh undefine "${IMAGE_KEY}"
+    fi
+    # Remove qcow2 file.
+    sudo rm -f "$LIBVIRT_IMAGE_PATH"
+    # Remomve tmp dir.
+    sudo rm -rf "$TEMPDIR"
+    # Stop httpd
+    sudo systemctl disable httpd --now
+}
+
+# Test result checking
+check_result () {
+    greenprint "Checking for test result"
+    if [[ $RESULTS == 1 ]]; then
+        greenprint "💚 Success"
+    else
+        redprint "❌ Failed"
+        clean_up
+        exit 1
+    fi
+}
+
+##################################################
+##
+## Upload ostree commit to pulp test
+##
+##################################################
+# Setup pulp server
+greenprint "📄 Setup pulp server with one container"
+mkdir -p settings pulp_storage pgsql containers
+echo "CONTENT_ORIGIN='http://$(hostname):8080'
+ANSIBLE_API_HOSTNAME='http://$(hostname):8080'
+ANSIBLE_CONTENT_HOSTNAME='http://$(hostname):8080/pulp/content'
+CACHE_ENABLED=True" >> settings/settings.py
+sudo podman run --detach \
+            --publish 9090:80 \
+            --name pulp \
+            --volume "$(pwd)/settings":/etc/pulp:Z \
+            --volume "$(pwd)/pulp_storage":/var/lib/pulp:Z \
+            --volume "$(pwd)/pgsql":/var/lib/pgsql:Z \
+            --volume "$(pwd)/containers":/var/lib/containers:Z \
+            --device /dev/fuse \
+            quay.io/pulp/pulp:nightly
+
+# Wait until pulp service is fully functional
+sleep 120
```

Comment:
Could you please use curl to confirm container running, not hard code wait?

Reply:
The sleep function here is not to wait container to be running, actually after the container is up and running, it still needs around 60 seconds to be full functional. 

For example, if run below command immediately after container is up and running,

podman exec -it pulp bash -c "pulpcore-manager reset-admin-password"

will get some internal errors, and wait around 60 seconds try again, it works well.